### PR TITLE
Sqlite connector

### PR DESCRIPTION
**Pull request by cbielow**
amendment to  #4280
which reverses the explicit ctor in `String` for `unsigned char*`.
Reasoning: its a bad place to put this functionality, since its only used in very few places (related to Sql) and std::string does not support this conversion either (and there is probably a very good reason for it). 

During refactoring, some other issues related to our Sql code popped up (e.g. potential mem-leaks), but thats for another PR (where we should make use of RAII to get rid of manual memory management for prepared statements, which currently require calling `sqlite3_finalize(stmt);` to free memory)

### CODE REVIEW

**Comment by hroest** [comment]
> (and there is probably a very good reason for it).

I pointed this out before, the reason is that `unsigned char*` are likely bytes of an unknown encoding (UTF8, UTF16, ???) and so converting that to std::string would mean you make implicit assumptions about your data for the gain of syntactic sugar instead of making this explicit. This will work here for SQL but maybe not in other places. Maybe a compromise would be to not have a constructor but have a static function `OpenMS::String::fromASCIIBytestream` where you are explicit that you are expecting only ASCII in your byte stream (maybe?) -- or your solution of doing this only for Sqlite.

File: src/openms/source/ANALYSIS/OPENSWATH/TransitionPQPFile.cpp [hunk 1]
```diff
@@ -244,17 +242,11 @@ namespace OpenMS
...
-      }
+      if (Sql::extractValue<std::string>(&tmp_field, stmt, 24)) tmp_field.split(';', mytransition.uniprot_id);
       Sql::extractValue<int>((int*)&mytransition.detecting_transition, stmt, 25);
       Sql::extractValue<int>((int*)&mytransition.identifying_transition, stmt, 26);
       Sql::extractValue<int>((int*)&mytransition.quantifying_transition, stmt, 27);
-      if (sqlite3_column_type( stmt, 28 ) != SQLITE_NULL)
-      {
-        String(reinterpret_cast<const char*>(sqlite3_column_text( stmt, 28 ))).split('|', mytransition.peptidoforms);
-      }
+      if (Sql::extractValue<std::string>(&tmp_field, stmt, 28)) tmp_field.split(';', mytransition.peptidoforms);
```
Comment (hroest) [inline src/openms/source/ANALYSIS/OPENSWATH/TransitionPQPFile.cpp]: copy-paste error
```suggestion
      if (Sql::extractValue<std::string>(&tmp_field, stmt, 28)) tmp_field.split('|', mytransition.peptidoforms);
```

Comment (cbielow) [reply src/openms/source/ANALYSIS/OPENSWATH/TransitionPQPFile.cpp]: whops. nice catch!

**Comment by cbielow** [comment]
rebuild jenkins

**Comment by jpfeuffer** [comment]
strcomp not declared in this scope. on gcc7. probably related.

**Comment by cbielow** [comment]
rebuild jenkins

File: src/openms/source/FORMAT/SqliteConnector.cpp [hunk 2]
```diff
@@ -131,26 +128,28 @@ namespace OpenMS
...
-      } 
+      }
     }
 
     rc = sqlite3_step(stmt);
     if (rc != SQLITE_DONE)
     {
       std::cerr << "SQL error after sqlite3_step" << std::endl;
       std::cerr << "Prepared statement " << prepare_statement << std::endl;
+      // TODO this is a mem-leak (missing sqlite3_finalize())
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/SqliteConnector.cpp]: Why not call it?

Comment (cbielow) [reply src/openms/source/FORMAT/SqliteConnector.cpp]: because that only fixes the problem locally. We should think about a more thorough solution?!

Comment (timosachsenberg) [reply src/openms/source/FORMAT/SqliteConnector.cpp]: Something like unique_ptr + lambda deleter?

Comment (cbielow) [reply src/openms/source/FORMAT/SqliteConnector.cpp]: yes. baked into `prepareStatement()` or some other place appropriate...

**Comment by cbielow** [comment]
MSVC has git issues, but since I've build it using VS15 im reasonably sure it works :)
will merge

File: src/openms/source/FORMAT/SqliteConnector.cpp (hunk 2 above)
Comment (hroest) [reply src/openms/source/FORMAT/SqliteConnector.cpp]: If you are here and throw an exception in the SQL code (e.g. invalid SQL), you probably have much bigger problems than a memory leak?

**Pull request merged**